File: net/crates/net/bindings/go/net/tool.go
// Go layer for AI tool calling on net.
//
// Wraps the existing TypedMeshRpc Go surface (TypedServe /
// TypedCallService) with the RegisterTool / CallTool ergonomic
// helpers + format translators that lower ToolDescriptor instances
// to OpenAI / Anthropic / MCP / Gemini tool shapes and parse
// provider tool-call replies back into nRPC dispatches.
//
// This is the Wave 3 / D-1 starting point. v1 covers unary
// register + invoke + format conversion. Streaming and discovery
// follow once the underlying CGO surface exposes them.
//
// Plan: see
// `crates/net/docs/plans/NRPC_AI_TOOL_CALLING_AND_AGENT_DX.md`,
// slices D-1 / D-2. Mirror of the Rust SDK's `net_sdk::tool` +
// `net_sdk::tool::formats`, the Node TS `tool.ts`, and the Python
// `net.tool` modules. Cross-language tests (T-1) will pin byte
// equality across all four implementations.

package net

import (
	"encoding/json"
	"errors"
	"fmt"
)

// =============================================================================
// Wire types — mirror of the Rust `ToolDescriptor` + `ToolEvent`.
// =============================================================================

// ToolDescriptor is the discovery shape for an AI tool, as
// advertised on the capability fold. One row per (ToolID, Version);
// NodeCount is filled by the aggregating walk (list_tools once it
// lands on the CGO surface).
//
// Wire-compatible 1:1 with the Rust substrate's `ToolDescriptor`,
// the Node TS `ToolDescriptor`, and the Python `ToolDescriptor`
// dataclass.
//
// Schemas are stored as JSON-encoded strings (matching the wire
// shape); use `json.Unmarshal([]byte(desc.InputSchema), &obj)` to
// get the parsed object for lowering into a provider tool
// definition.
type ToolDescriptor struct {
	ToolID          string   `json:"tool_id"`
	Name            string   `json:"name"`
	Version         string   `json:"version"`
	Description     string   `json:"description,omitempty"`
	InputSchema     string   `json:"input_schema,omitempty"`
	OutputSchema    string   `json:"output_schema,omitempty"`
	Requires        []string `json:"requires"`
	EstimatedTimeMs uint32   `json:"estimated_time_ms"`
	Stateless       bool     `json:"stateless"`
	Streaming       bool     `json:"streaming"`
	Tags            []string `json:"tags"`
	NodeCount       uint32   `json:"node_count"`
}

// ToolEventType discriminates ToolEvent variants. The wire JSON
// uses the `type` field; ToolEvent's MarshalJSON / UnmarshalJSON
// keeps the tagged-union shape symmetric with the other languages.
type ToolEventType string

const (
	ToolEventStart    ToolEventType = "start"
	ToolEventProgress ToolEventType = "progress"
	ToolEventDelta    ToolEventType = "delta"
	ToolEventResult   ToolEventType = "result"
	ToolEventError    ToolEventType = "error"
)

// ToolEvent is one envelope on a streaming tool. Wire-compatible
// 1:1 with the Rust `ToolEvent` enum: JSON-tagged with `type`,
// snake_case variants.
//
// Every stream ends with exactly one terminal event (Result or
// Error). Handlers that forget emit a synthesized
// `{type: "error", code: "missing_terminal", …}` from the Rust
// SDK's streaming wrapper.
type ToolEvent struct {
	Type ToolEventType `json:"type"`

	// Variant-specific fields — only the ones relevant to Type are
	// populated. Unmarshal sets the rest to zero values.

	// Start
	ToolID   string          `json:"tool_id,omitempty"`
	CallID   uint64          `json:"call_id,omitempty"`
	Metadata json.RawMessage `json:"metadata,omitempty"`

	// Progress
	Pct     *float32 `json:"pct,omitempty"`
	Message string   `json:"message,omitempty"`

	// Delta / Result
	Data json.RawMessage `json:"data,omitempty"`

	// Error
	Code    string          `json:"code,omitempty"`
	ErrMsg  string          `json:"-"` // Filled from `message` via UnmarshalJSON for Error variant.
	Details json.RawMessage `json:"details,omitempty"`
}

// IsTerminal returns true if the event closes the stream
// (Result or Error variant).
func (e ToolEvent) IsTerminal() bool {
	return e.Type == ToolEventResult || e.Type == ToolEventError
}

// =============================================================================
// Descriptor construction
// =============================================================================

// ToolOptions is the input to DescriptorFor / RegisterTool. Mirror
// of the Rust ToolMetadataBuilder, the Node TS ToolOptions, and the
// Python descriptor_for kwargs.
type ToolOptions struct {
	// Name is the nRPC service name + tool identifier. Required.
	Name string
	// Description is human-readable. Strongly recommended.
	Description string
	// Version defaults to "1.0.0".
	Version string
	// InputSchema is the JSON Schema for the request, as a Go
	// object (will be json.Marshal-ed to a string for storage).
	// Pass `nil` to omit.
	InputSchema interface{}
	// OutputSchema mirrors InputSchema.
	OutputSchema interface{}
	// Requires lists capability dependencies.
	Requires []string
	// EstimatedTimeMs is a soft latency hint (ms).
	EstimatedTimeMs uint32
	// Stateless defaults to true.
	Stateless *bool
	// Tags is a free-form list (e.g. ["web", "research"]).
	Tags []string
}

// DescriptorFor builds a ToolDescriptor from a ToolOptions literal.
// Schemas are json.Marshal-ed to strings; on encode failure (for
// non-serializable types) the field is left empty and the error is
// returned.
func DescriptorFor(opts ToolOptions) (ToolDescriptor, error) {
	desc := ToolDescriptor{
		ToolID:          opts.Name,
		Name:            opts.Name,
		Version:         opts.Version,
		Description:     opts.Description,
		Requires:        append([]string(nil), opts.Requires...),
		EstimatedTimeMs: opts.EstimatedTimeMs,
		Stateless:       true,
		Streaming:       false,
		Tags:            append([]string(nil), opts.Tags...),
		NodeCount:       0,
	}
	if desc.Version == "" {
		desc.Version = "1.0.0"
	}
	if opts.Stateless != nil {
		desc.Stateless = *opts.Stateless
	}
	if desc.Requires == nil {
		desc.Requires = []string{}
	}
	if desc.Tags == nil {
		desc.Tags = []string{}
	}
	if opts.InputSchema != nil {
		b, err := json.Marshal(opts.InputSchema)
		if err != nil {
			return desc, fmt.Errorf("descriptorFor: marshal input schema: %w", err)
		}
		desc.InputSchema = string(b)
	}
	if opts.OutputSchema != nil {
		b, err := json.Marshal(opts.OutputSchema)
		if err != nil {
			return desc, fmt.Errorf("descriptorFor: marshal output schema: %w", err)
		}
		desc.OutputSchema = string(b)
	}
	return desc, nil
}

// =============================================================================
// Register / invoke
// =============================================================================

// ToolServeHandle is returned by RegisterTool. Call Close() to
// deregister the underlying nRPC handler. Idempotent; second
// Close() is a no-op. Mirror of the Rust ToolServeHandle's Drop
// semantics.
//
// NOTE: v1 does NOT integrate with the substrate-side tool_registry,
// so the `ai-tool:<tool_id>` capability tag must be added to the
// caller's announce explicitly. See AddToolCapabilitiesToAnnounce.
// Once the CGO surface exposes tool_registry (Wave 3 follow-up),
// this handle will atomically reverse both the registry insert and
// the handler registration on Close().
type ToolServeHandle struct {
	Descriptor ToolDescriptor
	inner      *ServeHandle
	closed     bool
}

// Close deregisters the handler. Idempotent.
func (h *ToolServeHandle) Close() {
	if h.closed {
		return
	}
	h.closed = true
	if h.inner != nil {
		h.inner.Close()
	}
}

// RegisterTool registers a typed handler as an AI tool against
// `rpc`. The handler is registered as an nRPC service at
// `descriptor.ToolID` with JSON codec.
//
// The caller is responsible for announcing the tool to peers — use
// AddToolCapabilitiesToAnnounce on the CapabilitySetWire passed to
// the mesh's announce surface so the `ai-tool:<tool_id>` tag lands
// on the wire.
//
// Wave 3 follow-up: once the CGO surface exposes tool_registry(),
// this helper will atomically insert there too, making the
// announce-time merge automatic (matching the Rust SDK's contract).
func RegisterTool[Req, Resp any](
	rpc *TypedMeshRpc,
	descriptor ToolDescriptor,
	handler TypedHandler[Req, Resp],
) (*ToolServeHandle, error) {
	inner, err := TypedServe[Req, Resp](rpc, descriptor.ToolID, handler)
	if err != nil {
		return nil, err
	}
	return &ToolServeHandle{Descriptor: descriptor, inner: inner}, nil
}

// CallTool dispatches a capability-routed unary tool invocation
// via TypedCallService. JSON codec is hardwired — every AI
// provider (OpenAI, Anthropic, Gemini, MCP) consumes JSON for tool
// input/output.
//
// Returns the decoded response, or an error (NoRoute if no host
// advertises `nrpc:<toolID>`, bubbled handler errors otherwise).
func CallTool[Req, Resp any](
	rpc *TypedMeshRpc,
	toolID string,
	request Req,
) (Resp, error) {
	return TypedCallService[Req, Resp](rpc, toolID, request, nil)
}

// AddToolCapabilitiesToAnnounce merges tool descriptors into a
// CapabilitySetWire so the next announce carries:
//
//   - `ai-tool:<tool_id>` tag — peer fold's tag-prefix lookup hits.
//   - The `tool::<id>::input_schema` / `output_schema` metadata
//     keys for fold-side hydration.
//
// Returns the same wire object for chaining. v1 convenience; once
// the CGO surface exposes tool_registry, this becomes optional.
func AddToolCapabilitiesToAnnounce(
	caps CapabilitySetWire,
	descriptors []ToolDescriptor,
) CapabilitySetWire {
	if len(descriptors) == 0 {
		return caps
	}
	// Dedupe tags.
	tagSet := make(map[string]struct{}, len(caps.Tags))
	for _, t := range caps.Tags {
		tagSet[t] = struct{}{}
	}
	if caps.Metadata == nil {
		caps.Metadata = make(map[string]string)
	}
	for _, desc := range descriptors {
		tag := "ai-tool:" + desc.ToolID
		if _, ok := tagSet[tag]; !ok {
			tagSet[tag] = struct{}{}
			caps.Tags = append(caps.Tags, tag)
		}
		// Mirror schemas into metadata under the substrate's
		// canonical key shape (matches
		// ToolCapability::input_schema_metadata_key in Rust).
		if desc.InputSchema != "" {
			caps.Metadata["tool::"+desc.ToolID+"::input_schema"] = desc.InputSchema
		}
		if desc.OutputSchema != "" {
			caps.Metadata["tool::"+desc.ToolID+"::output_schema"] = desc.OutputSchema
		}
		if desc.Description != "" {
			caps.Metadata["tool::"+desc.ToolID+"::description"] = desc.Description
		}
		if desc.Streaming {
			caps.Metadata["tool::"+desc.ToolID+"::streaming"] = "1"
		}
		if len(desc.Tags) > 0 {
			// Comma-joined (substrate convention).
			joined := ""
			for i, t := range desc.Tags {
				if i > 0 {
					joined += ","
				}
				joined += t
			}
			caps.Metadata["tool::"+desc.ToolID+"::tags"] = joined
		}
	}
	return caps
}

// =============================================================================
// Format translators — mirror of `net_sdk::tool::formats`
// =============================================================================

// ToolCallSpec is the canonical hand-off between an LLM-provider
// adapter and CallTool. ArgumentsJSON is a string so the boundary
// is provider-agnostic (OpenAI's arguments arrive as a string
// anyway; Anthropic/MCP/Gemini's parsed objects re-serialize once).
type ToolCallSpec struct {
	Name              string
	ArgumentsJSON     string
	ProviderCallID    string
	HasProviderCallID bool
}

// ToolCallParseError is returned when a provider's tool-call reply
// doesn't match its spec.
var ErrToolCallParse = errors.New("net.tool: provider tool-call reply parse error")

func parseError(msg string) error {
	return fmt.Errorf("%w: %s", ErrToolCallParse, msg)
}

// inputSchemaValue parses desc.InputSchema into a Go value, falling
// back to `{"type":"object","properties":{}}` if missing/malformed.
// Providers' strict-mode validators reject null parameter schemas.
func inputSchemaValue(desc ToolDescriptor) interface{} {
	if desc.InputSchema == "" {
		return map[string]interface{}{"type": "object", "properties": map[string]interface{}{}}
	}
	var v interface{}
	if err := json.Unmarshal([]byte(desc.InputSchema), &v); err != nil {
		return map[string]interface{}{"type": "object", "properties": map[string]interface{}{}}
	}
	return v
}

// ---- OpenAI ----

// ToOpenAITool lowers a descriptor to an OpenAI tool definition:
//
//	{type: "function", function: {name, description, parameters, strict}}
//
// `strict` is true when the descriptor carried an InputSchema (i.e.
// publishable on the fold).
func ToOpenAITool(desc ToolDescriptor) map[string]interface{} {
	return map[string]interface{}{
		"type": "function",
		"function": map[string]interface{}{
			"name":        desc.ToolID,
			"description": desc.Description,
			"parameters":  inputSchemaValue(desc),
			"strict":      desc.InputSchema != "",
		},
	}
}

// LowerOpenAIToolCall parses one OpenAI `tool_calls[]` entry into
// a ToolCallSpec. OpenAI's `function.arguments` is a JSON-encoded
// STRING; validates the string parses up front to fail fast.
func LowerOpenAIToolCall(call map[string]interface{}) (ToolCallSpec, error) {
	spec := ToolCallSpec{}
	fn, ok := call["function"].(map[string]interface{})
	if !ok {
		return spec, parseError("tool-call reply missing field `function`")
	}
	name, ok := fn["name"].(string)
	if !ok {
		return spec, parseError("tool-call reply field `function.name` must be a string")
	}
	args, ok := fn["arguments"].(string)
	if !ok {
		return spec, parseError("tool-call reply field `function.arguments` must be a JSON-encoded string")
	}
	var probe interface{}
	if err := json.Unmarshal([]byte(args), &probe); err != nil {
		return spec, parseError(fmt.Sprintf("tool-call arguments were not valid JSON: %v", err))
	}
	spec.Name = name
	spec.ArgumentsJSON = args
	if id, ok := call["id"].(string); ok {
		spec.ProviderCallID = id
		spec.HasProviderCallID = true
	}
	return spec, nil
}

// ---- Anthropic ----

// ToAnthropicTool lowers a descriptor to an Anthropic tool
// definition: {name, description, input_schema}. Anthropic has no
// tool-level `strict` flag.
func ToAnthropicTool(desc ToolDescriptor) map[string]interface{} {
	return map[string]interface{}{
		"name":         desc.ToolID,
		"description":  desc.Description,
		"input_schema": inputSchemaValue(desc),
	}
}

// LowerAnthropicToolUse parses one Anthropic `tool_use` content
// block. `input` is already a parsed object; re-serializes once to
// preserve the `ArgumentsJSON: string` invariant.
func LowerAnthropicToolUse(block map[string]interface{}) (ToolCallSpec, error) {
	spec := ToolCallSpec{}
	name, ok := block["name"].(string)
	if !ok {
		return spec, parseError("tool_use block field `name` must be a string")
	}
	input, exists := block["input"]
	if !exists {
		return spec, parseError("tool_use block missing field `input`")
	}
	b, err := json.Marshal(input)
	if err != nil {
		return spec, parseError(fmt.Sprintf("tool_use input re-serialize failed: %v", err))
	}
	spec.Name = name
	spec.ArgumentsJSON = string(b)
	if id, ok := block["id"].(string); ok {
		spec.ProviderCallID = id
		spec.HasProviderCallID = true
	}
	return spec, nil
}

// ---- MCP ----

// ToMCPTool lowers a descriptor to an MCP tool definition:
// {name, description, inputSchema} (camelCase).
func ToMCPTool(desc ToolDescriptor) map[string]interface{} {
	return map[string]interface{}{
		"name":        desc.ToolID,
		"description": desc.Description,
		"inputSchema": inputSchemaValue(desc),
	}
}

// LowerMCPToolsCall parses an MCP `tools/call` request's `params`
// into a ToolCallSpec. ProviderCallID is left unset — MCP's
// JSON-RPC `id` lives one envelope layer up.
func LowerMCPToolsCall(params map[string]interface{}) (ToolCallSpec, error) {
	spec := ToolCallSpec{}
	name, ok := params["name"].(string)
	if !ok {
		return spec, parseError("tools/call params field `name` must be a string")
	}
	args, exists := params["arguments"]
	if !exists {
		return spec, parseError("tools/call params missing field `arguments`")
	}
	b, err := json.Marshal(args)
	if err != nil {
		return spec, parseError(fmt.Sprintf("tools/call arguments re-serialize failed: %v", err))
	}
	spec.Name = name
	spec.ArgumentsJSON = string(b)
	return spec, nil
}

// ---- Gemini ----

// ToGeminiFunctionDeclaration lowers a descriptor to one Gemini
// FunctionDeclaration: {name, description, parameters}. Caller
// wraps these into the outer
// `tools: [{ function_declarations: [...] }]` array.
func ToGeminiFunctionDeclaration(desc ToolDescriptor) map[string]interface{} {
	return map[string]interface{}{
		"name":        desc.ToolID,
		"description": desc.Description,
		"parameters":  inputSchemaValue(desc),
	}
}

// LowerGeminiFunctionCall parses one Gemini `functionCall` part.
// Gemini has no per-call id; ProviderCallID is left unset.
func LowerGeminiFunctionCall(call map[string]interface{}) (ToolCallSpec, error) {
	spec := ToolCallSpec{}
	name, ok := call["name"].(string)
	if !ok {
		return spec, parseError("functionCall field `name` must be a string")
	}
	args, exists := call["args"]
	if !exists {
		return spec, parseError("functionCall missing field `args`")
	}
	b, err := json.Marshal(args)
	if err != nil {
		return spec, parseError(fmt.Sprintf("functionCall args re-serialize failed: %v", err))
	}
	spec.Name = name
	spec.ArgumentsJSON = string(b)
	return spec, nil
}

File: net/crates/net/bindings/node/tool.ts
// TypeScript layer for AI tool calling on net.
//
// Wraps the existing `TypedMeshRpc` napi surface with the `tool()` /
// `callTool()` ergonomic helpers + format translators that lower
// `ToolDescriptor`s to OpenAI / Anthropic / MCP / Gemini tool shapes
// and parse provider tool-call replies back into nRPC dispatches.
//
// This is the Wave 3 / B-1 + B-4 starting point. v1 covers unary
// register + invoke + format conversion. Streaming (B-2) and
// discovery (B-3 list_tools / watch_tools) follow once the
// underlying napi surface exposes them; today the only available
// streaming primitive is direct-addressed (`callStreaming(nodeId,
// ...)`), so capability-routed streaming has to wait on a
// `callServiceStreaming` TS wrapper or a `findServiceNodes` +
// direct-call composition (TODO).
//
// Plan: see
// `crates/net/docs/plans/NRPC_AI_TOOL_CALLING_AND_AGENT_DX.md`,
// slices B-1 / B-2 / B-4. Mirror of the Rust SDK's
// `net_sdk::tool` + `net_sdk::tool::formats` modules — cross-
// language tests (T-1) will pin byte equality across both.

import type { CallOptions, ServeHandle, TypedMeshRpc } from './mesh_rpc'
import type { CapabilitySetJs, ToolJs } from './index'

// ============================================================================
// Wire types — mirror of the Rust `ToolDescriptor` + `ToolEvent`.
// ============================================================================

/**
 * Discovery shape for an AI tool, as advertised on the capability
 * fold. One row per `(tool_id, version)`; `nodeCount` is filled by
 * the aggregating walk (`list_tools` once it lands).
 *
 * Schemas are stored as JSON-encoded strings (matching the Rust
 * substrate's wire shape). Use `JSON.parse(desc.inputSchema)` to
 * get the parsed JSON Schema object — most consumers want this
 * for lowering into a provider tool definition.
 *
 * Wire-compatible 1:1 with `net::adapter::net::cortex::tool::ToolDescriptor`.
 */
export interface ToolDescriptor {
  /** nRPC service name. Same string `callTool(...)` takes. */
  toolId: string
  /** Human-readable name. Defaults to `toolId` if unset. */
  name: string
  /** Tool version (semver-ish). Defaults to `"1.0.0"`. */
  version: string
  /** Human-readable description; LLMs read this to decide when to call. */
  description?: string
  /** JSON-encoded JSON Schema (draft 2020-12) for the request body. */
  inputSchema?: string
  /** JSON-encoded JSON Schema (draft 2020-12) for the response body. */
  outputSchema?: string
  /** Required capabilities / dependencies (free-form strings). */
  requires: string[]
  /** Soft latency hint (ms); `0` = no estimate. */
  estimatedTimeMs: number
  /** True if the tool is a pure function (no session state). */
  stateless: boolean
  /** True if the tool is server-streaming (uses `serveToolStreaming`). */
  streaming: boolean
  /** Free-form host-attached tags (e.g. `["web", "research"]`). */
  tags: string[]
  /** How many nodes currently serve this `(toolId, version)`. */
  nodeCount: number
}

/**
 * One envelope on a streaming tool. Discriminated by `type`.
 *
 * Wire-compatible 1:1 with `net::adapter::net::cortex::tool::ToolEvent`
 * (JSON tag form, `{"type": "start", …}` shape).
 *
 * Every stream ends with exactly one terminal event
 * (`type: "result"` or `type: "error"`). Handlers that forget emit
 * a synthesized `{type: "error", code: "missing_terminal", …}` from
 * the Rust SDK's streaming wrapper.
 */
export type ToolEvent =
  | ToolEventStart
  | ToolEventProgress
  | ToolEventDelta
  | ToolEventResult
  | ToolEventError

export interface ToolEventStart {
  type: 'start'
  toolId: string
  callId?: number
  metadata?: unknown
}

export interface ToolEventProgress {
  type: 'progress'
  pct?: number
  message?: string
}

export interface ToolEventDelta {
  type: 'delta'
  data: unknown
}

export interface ToolEventResult {
  type: 'result'
  data: unknown
}

export interface ToolEventError {
  type: 'error'
  code: string
  message: string
  details?: unknown
}

/** True if `event` is a terminal envelope (`result` or `error`). */
export function isTerminalEvent(event: ToolEvent): boolean {
  return event.type === 'result' || event.type === 'error'
}

// ============================================================================
// Descriptor construction
// ============================================================================

/**
 * Options for `tool({...})` and `serveTool(rpc, options, handler)`.
 * Mirror of the Rust `ToolMetadataBuilder` shape — caller supplies
 * the fields that don't derive from a type signature in JS (no
 * compile-time type system to introspect, unlike `schemars` in Rust).
 *
 * `inputSchema` / `outputSchema` are JSON-Schema-as-object (caller
 * uses `zod-to-json-schema`, `pydantic`, or hand-rolls); we serialize
 * to a string before stashing on the descriptor.
 */
export interface ToolOptions {
  /** nRPC service name + tool identifier. Required. */
  name: string
  /** Human-readable description. Strongly recommended. */
  description?: string
  /** Version. Defaults to `"1.0.0"`. */
  version?: string
  /** JSON Schema object for the request. */
  inputSchema?: object
  /** JSON Schema object for the response. */
  outputSchema?: object
  /** Required capabilities / dependencies. */
  requires?: string[]
  /** Soft latency hint (ms). */
  estimatedTimeMs?: number
  /** Pure-function flag. Default `true`. */
  stateless?: boolean
  /** Free-form tags. */
  tags?: string[]
}

/** Construct a [`ToolDescriptor`] from a `ToolOptions` literal. */
export function descriptorFrom(options: ToolOptions): ToolDescriptor {
  return {
    toolId: options.name,
    name: options.name,
    version: options.version ?? '1.0.0',
    description: options.description,
    inputSchema: options.inputSchema ? JSON.stringify(options.inputSchema) : undefined,
    outputSchema: options.outputSchema ? JSON.stringify(options.outputSchema) : undefined,
    requires: options.requires ?? [],
    estimatedTimeMs: options.estimatedTimeMs ?? 0,
    stateless: options.stateless ?? true,
    streaming: false,
    tags: options.tags ?? [],
    nodeCount: 0,
  }
}

// ============================================================================
// Register / invoke
// ============================================================================

/**
 * Handler signature for `serveTool` — receives a decoded request,
 * returns a decoded response (or a Promise of one).
 */
export type ToolHandler<Req = unknown, Resp = unknown> = (
  req: Req,
) => Resp | Promise<Resp>

/**
 * Handle returned by `serveTool`. Calling `.close()` deregisters the
 * underlying nRPC handler (mirror of the Rust `ToolServeHandle`'s
 * Drop semantics). Calling `.close()` twice is idempotent; the
 * second call is a no-op.
 *
 * NOTE: v1 does NOT yet integrate with the substrate-side
 * `tool_registry`, so the `ai-tool:<toolId>` capability tag must be
 * added to the caller's announce explicitly. See
 * [`addToolCapabilitiesToAnnounce`] for the convention. Once the
 * napi surface exposes `tool_registry()` insert/remove (a Wave 3
 * follow-up), this handle will atomically reverse both the
 * registry insert and the handler registration on `.close()`.
 */
export interface ToolServeHandle {
  /** The descriptor under which the tool was registered. */
  readonly descriptor: ToolDescriptor
  /** Deregister the handler. Idempotent. */
  close(): void
}

/**
 * Register an AI tool against `rpc`. The handler is registered as
 * an nRPC service at `descriptor.toolId` with JSON codec (same as
 * the Rust SDK's `Mesh::serve_tool`).
 *
 * The caller is responsible for announcing the tool to peers — use
 * [`addToolCapabilitiesToAnnounce`] on the `CapabilitySetJs` you
 * pass to `mesh.announceCapabilities(...)` so the
 * `ai-tool:<toolId>` tag + the `ToolJs` entry land on the wire.
 *
 * Wave 3 follow-up: once the napi surface exposes
 * `tool_registry()`, this helper will atomically insert there too,
 * making the announce-time merge automatic (matching the Rust
 * SDK's contract).
 */
export function serveTool<Req = unknown, Resp = unknown>(
  rpc: TypedMeshRpc,
  options: ToolOptions,
  handler: ToolHandler<Req, Resp>,
): ToolServeHandle {
  const descriptor = descriptorFrom(options)
  const inner: ServeHandle = rpc.serve<Req, Resp>(descriptor.toolId, handler)
  let closed = false
  return {
    descriptor,
    close() {
      if (closed) return
      closed = true
      inner.close()
    },
  }
}

/**
 * Capability-routed unary tool invocation. Encodes `req` as JSON
 * (the codec every AI provider consumes for tool input/output),
 * dispatches via `rpc.callService(toolId, req, opts)`.
 *
 * Throws `NoRouteError` if no host advertises `nrpc:<toolId>` in
 * the local capability fold; bubbles handler errors as
 * `RpcServerError` with the typed-handler status code.
 */
export async function callTool<Req = unknown, Resp = unknown>(
  rpc: TypedMeshRpc,
  toolId: string,
  req: Req,
  opts?: CallOptions,
): Promise<Resp> {
  return rpc.callService<Req, Resp>(toolId, req, opts)
}

/**
 * Merge tool descriptors into a `CapabilitySetJs` so the next
 * `mesh.announceCapabilities(caps)` carries:
 *
 * - `ai-tool:<toolId>` tag — peer fold's tag-prefix lookup hits.
 * - A `ToolJs` entry — peer's `list_tools` walk sees the
 *   tool's tag-encoded fields.
 *
 * Caller still owns the `caps` object — pass it through
 * `mesh.announceCapabilities(caps)` to publish. Returns the same
 * object for chaining.
 *
 * This is a v1 convenience; once the napi surface exposes
 * `tool_registry()`, the announce-time merge happens
 * automatically and this helper becomes optional.
 */
export function addToolCapabilitiesToAnnounce(
  caps: CapabilitySetJs,
  descriptors: ToolDescriptor[],
): CapabilitySetJs {
  if (descriptors.length === 0) return caps
  const tags = new Set(caps.tags ?? [])
  const tools: ToolJs[] = [...(caps.tools ?? [])]
  for (const desc of descriptors) {
    tags.add(`ai-tool:${desc.toolId}`)
    tools.push({
      toolId: desc.toolId,
      name: desc.name,
      version: desc.version,
      inputSchema: desc.inputSchema,
      outputSchema: desc.outputSchema,
      requires: desc.requires,
      estimatedTimeMs: desc.estimatedTimeMs,
      stateless: desc.stateless,
    })
  }
  caps.tags = Array.from(tags)
  caps.tools = tools
  return caps
}

// ============================================================================
// Format translators — mirror of `net_sdk::tool::formats`
// ============================================================================
//
// Each provider submodule exports two directions:
//
// 1. `to<Provider>Tool(desc) -> object` — descriptor → provider's
//    tool-definition shape for the `tools` array on the provider's
//    HTTP request.
// 2. `lower<Provider>ToolCall(reply) -> ToolCallSpec` — provider's
//    reply → ToolCallSpec the caller hands to `callTool`.
//
// All translators short-circuit a missing `inputSchema` to an
// empty-object schema (`{type: "object", properties: {}}`) since
// providers' strict-mode validators reject null parameter schemas.

/** Canonical hand-off between a provider adapter and `callTool`. */
export interface ToolCallSpec {
  /** nRPC tool_id to invoke. */
  name: string
  /** JSON-encoded arguments to pass to `callTool` (caller parses). */
  argumentsJson: string
  /** Provider-supplied call id when present (for reply correlation). */
  providerCallId?: string
}

/** Thrown when a provider's tool-call reply doesn't match its spec. */
export class ToolCallParseError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ToolCallParseError'
  }
}

function inputSchemaValue(desc: ToolDescriptor): object {
  if (!desc.inputSchema) return { type: 'object', properties: {} }
  try {
    return JSON.parse(desc.inputSchema) as object
  } catch {
    // Schema string was malformed (shouldn't happen for descriptors
    // built via `descriptorFrom`). Empty-object fallback keeps
    // provider validators happy.
    return { type: 'object', properties: {} }
  }
}

/** OpenAI Chat Completions / Responses API `tools` array. */
export const openai = {
  /**
   * Lower a descriptor to an OpenAI tool definition. Shape:
   * ```
   * { type: "function", function: { name, description, parameters, strict } }
   * ```
   * `strict` is true when the descriptor carried an `inputSchema`.
   */
  toOpenaiTool(desc: ToolDescriptor): object {
    return {
      type: 'function',
      function: {
        name: desc.toolId,
        description: desc.description ?? '',
        parameters: inputSchemaValue(desc),
        strict: desc.inputSchema !== undefined,
      },
    }
  },

  /**
   * Parse one OpenAI `tool_calls[]` entry into a `ToolCallSpec`.
   * OpenAI's `function.arguments` is a JSON-encoded STRING; this
   * helper validates it parses up front so malformed payloads fail
   * fast instead of riding through `callTool`.
   */
  lowerOpenaiToolCall(call: Record<string, unknown>): ToolCallSpec {
    const fn = call['function'] as Record<string, unknown> | undefined
    if (!fn) throw new ToolCallParseError('tool-call reply missing field `function`')
    const name = fn['name']
    if (typeof name !== 'string') {
      throw new ToolCallParseError('tool-call reply field `function.name` must be a string')
    }
    const argumentsField = fn['arguments']
    if (typeof argumentsField !== 'string') {
      throw new ToolCallParseError(
        'tool-call reply field `function.arguments` must be a JSON-encoded string',
      )
    }
    try {
      JSON.parse(argumentsField)
    } catch (e) {
      throw new ToolCallParseError(
        `tool-call arguments were not valid JSON: ${(e as Error).message}`,
      )
    }
    const id = call['id']
    return {
      name,
      argumentsJson: argumentsField,
      providerCallId: typeof id === 'string' ? id : undefined,
    }
  },
}

/** Anthropic Messages API `tools` array + `tool_use` content blocks. */
export const anthropic = {
  /**
   * Lower a descriptor to an Anthropic tool definition. Shape:
   * ```
   * { name, description, input_schema }
   * ```
   * No tool-level `strict` flag — Anthropic relies on schema-
   * validated tool input as the default.
   */
  toAnthropicTool(desc: ToolDescriptor): object {
    return {
      name: desc.toolId,
      description: desc.description ?? '',
      input_schema: inputSchemaValue(desc),
    }
  },

  /**
   * Parse one Anthropic `tool_use` content block into a
   * `ToolCallSpec`. `input` is already a parsed object (not a
   * string like OpenAI); re-serializes once to preserve the
   * `argumentsJson: string` invariant.
   */
  lowerAnthropicToolUse(block: Record<string, unknown>): ToolCallSpec {
    const name = block['name']
    if (typeof name !== 'string') {
      throw new ToolCallParseError('tool_use block field `name` must be a string')
    }
    if (!('input' in block)) {
      throw new ToolCallParseError('tool_use block missing field `input`')
    }
    const argumentsJson = JSON.stringify(block['input'])
    const id = block['id']
    return {
      name,
      argumentsJson,
      providerCallId: typeof id === 'string' ? id : undefined,
    }
  },
}

/** Model Context Protocol `tools/list` + `tools/call`. */
export const mcp = {
  /** Lower a descriptor to an MCP tool definition. Shape: `{ name, description, inputSchema }` (camelCase). */
  toMcpTool(desc: ToolDescriptor): object {
    return {
      name: desc.toolId,
      description: desc.description ?? '',
      inputSchema: inputSchemaValue(desc),
    }
  },

  /**
   * Parse an MCP `tools/call` request's `params` into a
   * `ToolCallSpec`. `providerCallId` is left `undefined` — MCP's
   * JSON-RPC `id` lives one envelope layer up, threaded
   * independently.
   */
  lowerMcpToolsCall(params: Record<string, unknown>): ToolCallSpec {
    const name = params['name']
    if (typeof name !== 'string') {
      throw new ToolCallParseError('tools/call params field `name` must be a string')
    }
    if (!('arguments' in params)) {
      throw new ToolCallParseError('tools/call params missing field `arguments`')
    }
    return {
      name,
      argumentsJson: JSON.stringify(params['arguments']),
      providerCallId: undefined,
    }
  },
}

/** Gemini `generateContent` function-calling shape. */
export const gemini = {
  /**
   * Lower a descriptor to one Gemini `FunctionDeclaration`. Shape:
   * ```
   * { name, description, parameters }
   * ```
   * Caller wraps these into the outer
   * `tools: [{ function_declarations: [ … ] }]` array.
   */
  toGeminiFunctionDeclaration(desc: ToolDescriptor): object {
    return {
      name: desc.toolId,
      description: desc.description ?? '',
      parameters: inputSchemaValue(desc),
    }
  },

  /**
   * Parse one Gemini `functionCall` part into a `ToolCallSpec`.
   * Gemini has no per-call id; the spec leaves `providerCallId`
   * `undefined` (multi-call sequences are positional).
   */
  lowerGeminiFunctionCall(call: Record<string, unknown>): ToolCallSpec {
    const name = call['name']
    if (typeof name !== 'string') {
      throw new ToolCallParseError('functionCall field `name` must be a string')
    }
    if (!('args' in call)) {
      throw new ToolCallParseError('functionCall missing field `args`')
    }
    return {
      name,
      argumentsJson: JSON.stringify(call['args']),
      providerCallId: undefined,
    }
  },
}
